src/cuda/ptx.cu

```diff
@@ -85,6 +85,12 @@ extern "C" int sloth256_189_cuda_batch_encode(unsigned int piece[], size_t len,
             // have any remainder
             block_count = (remaining_piece_size / 4096) / thread_count;
 
+            // however, we might need less than 256 threads
```

Comment:
Our public API protects the invariant where number of pieces is always multiple of 1024. Isn't that sufficient?

Also now I'm wondering why we need a multiple of 1024 if code apparently works with batches of 256.

Reply:
Unfortunately it is not enough. In case of GPU being utilized by another program, in each round we are trying to allocate memory in the gpu, and if that fails, we try to allocate half of the memory, and so on... so it could be that the remaining pieces for the next round may require less than 256 thread. 

Each thread is responsible from 1 piece and per one block we are using 256 thread for efficiency. Each piece is 4kb, hence the size of 1024

Reply:
So `block_count` is actually `piece_count` in reality, is that correct?

In terms of memory allocation, why don't you allocate once outside of the loop and always know how much memory you have available so you don't have to make these check on every iteration of the loop?

Reply:
`block_count` is not `piece_count`. We are assigning a piece per thread, and default is 256 threads per block.

Btw I will implement the suggestion of allocating memory once, and copying over it in each iteration, that will simplify the code significantly.

Reply:
The suggestion is implemented. 
Also, now it starts to allocate the default memory proportional to the length of the input data. So there won't be any leftovers, and each round will chunk the input data in equal sizes. So this check is not necessary anymore :)